Silent Payments: Implement BIP352

josibake opened the pull request:
This PR is part of integrating silent payments into Bitcoin Core. The project is tracked in https://github.com/bitcoin/bitcoin/issues/28536

Based on https://github.com/bitcoin-core/secp256k1/pull/1698

## BIP352

This PR focuses strictly on the BIP logic and attempts to separate it from the wallet and transaction implementation details. This is accomplished by working directly with public and private keys, instead of needing a wallet backend and transactions for testing. Labels for the receiver are optional and thus deferred for a later PR.

Test vectors from the BIP are included as unit tests.

josibake committed:
Add "sp" HRP

theStack commented on src/kernel/chainparams.cpp:
```diff
@@ -508,6 +511,7 @@ class CRegTestParams : public CChainParams
         base58Prefixes[EXT_SECRET_KEY] = {0x04, 0x35, 0x83, 0x94};
 
         bech32_hrp = "bcrt";
+        silent_payment_hrp = "sprt";
```
Didn't verify my theory in practice yet, but according to my calculations having a HRP of size 4 would exceed the current 117 character limit (see https://github.com/bitcoin/bips/pull/1458/files#r1271312933). I assume the regtest-HRP should just also be set to "tsp" here (as it's currently stated in the BIP)? Otherweise the limit has to be raised by one.

Fun fact: in #27827 the "sprt" HRP works, but this is because the pubkeys are encoded as x-only (see commit https://github.com/bitcoin/bitcoin/pull/27827/commits/3a0d081ee07f3c65220a45855c96bf35ac2643fd), i.e. the payload of the address is only 64 bytes instead of 66 bytes and hence the limit is not exceeded.

josibake replied:
I've updated the BIP to recommend upping the character limit to 1023, as having a limit of 117 would cause issues with forward compatibility with future silent payment addresses. I've also removed regtest from the BIP, as I don't think it's correct to define Bitcoin Core specific conventions in BIPs. Regarding "tsp" vs "sprt", I don't really have a strong opinion here, but would prefer to follow the convention in Bitcoin Core. If we end up moving Bitcoin Core to use the same HRP for all networks, then we should update the silent payments HRP to do the same.

achow101 commented on src/key.cpp:
```diff
@@ -172,6 +174,79 @@ bool CKey::Negate()
+{
+    assert(fValid);
+
+    unsigned char tweaked_seckey[32];
+    memcpy(tweaked_seckey, data(), 32);
```
In 56882622faf469b6f948f79a69c3c8ddbde92ff8 "crypto: update CKey, CPubkey for silent payments"

We shouldn't be copying secret data out of a secure structure into an insecure one. When `tweaked_seckey` is destroyed, the underlying data may still linger in memory and end up being read by something else.

If this function must return a new `CKey`, then it should instantiate it here and perform the operation on it's `data()` rather than doing this copy.

However, I would prefer if this function mirrored the libsecp one where the tweaking occurs on and modified this `CKey` itself.

josibake replied:
Now operates on `keydata.data()` in place

achow101 commented on src/key.cpp:
```diff
@@ -172,6 +174,79 @@ bool CKey::Negate()
+{
+    assert(fValid);
+
+    unsigned char tweaked_seckey[32];
+    memcpy(tweaked_seckey, data(), 32);
```
In 56882622faf469b6f948f79a69c3c8ddbde92ff8 "crypto: update CKey, CPubkey for silent payments"

Same data copying concerns as `AddTweak`.

josibake replied:
Fixed

aureleoules reviewed (commented):
Left some comments. I will review further later.

josibake commented:
> Needs rebase, if still relevant

thanks, fixed!

josibake commented on src/kernel/chainparams.cpp:
```diff
@@ -508,6 +511,7 @@ class CRegTestParams : public CChainParams
         base58Prefixes[EXT_SECRET_KEY] = {0x04, 0x35, 0x83, 0x94};
 
         bech32_hrp = "bcrt";
+        silent_payment_hrp = "sprt";
```
This issue has some helpful context: https://github.com/bitcoin/bitcoin/issues/12314#issuecomment-582600899

TLDR; at the time regtest was created, it was assumed that using a separate HRP wouldn't cost anything.

> If each of the four had a different one, I’d get it, but why does regtest have a different one than the other two testing networks?

This actually makes sense to me. Testnet and Signet are "global" networks, so everyone has to agree on the address format. Regtest is always used in a private context, and is specific to Bitcoin Core. Having a separate HRP helps differentiate the local vs networked usage IMO.

> Unless I’m forgetting about some good reason for that to be there, I might recommend that you just use tsp for all three

If we decided to make the HRP for regtest throughout Bitcoin Core to be the same as signet and testnet, then I'm fine with this. What I don't want is for silent payment addresses to be a special case in the functional test framework.

Sjors replied:
As long as testnet and signet are the same, and different from mainnet, I'm happy. When testing with e.g. hardware wallets it lets you use the testnet app / firmware instead of signet-specific stuff.

murchandamus replied:
@josibake: It’s rather that bech32(m) addresses are the odd ones out. Before native segwit outputs all addresses across testnet and regtest were the same. After rereading #12314, I would still recommend that you use the same hrp for all test-networks.

josibake replied:
In the interest of keeping this PR well scoped, I think we should go for consistency. Since silent payment addresses are bech32m encoded, we should follow the convention _currently_ used for bech32(m), and if we do decided to change the convention for bech32(m) hrp's in the future, then it should change for all of them (including silent payments)

josibake commented on src/bech32.cpp:
```diff
@@ -370,11 +370,13 @@ std::string Encode(Encoding encoding, const std::string& hrp, const data& values
 /** Decode a Bech32 or Bech32m string. */
-DecodeResult Decode(const std::string& str) {
+DecodeResult Decode(const std::string& str, bool silent) {
+    std::size_t max_size = silent ? 1024 : 90;
```
Note to self, this should be 1023

josibake replied:
Wrap this function in a more generic "Decode" function that doesn't have size restrictions, possibly use an ENUM for indicating what sort of error guarantees you want.

BrandonOdiwuor commented on src/key_io.cpp:
```diff
@@ -309,3 +311,51 @@ bool IsValidDestinationString(const std::string& str)
+    if (!ConvertBits<5, 8, false>([&](unsigned char c) { silent_payment_data.push_back(c); }, dec.data.begin() + 1, dec.data.end())) {
+        return {};
+    }
+    if ((version == 0 && silent_payment_data.size() != SILENT_PAYMENT_V0_DATA_SIZE) || silent_payment_data.size() < SILENT_PAYMENT_V0_DATA_SIZE) return {};
```
To improve maintainability and support forward compatibility, consider abstracting the handling of different versions into a separate helper function like ‘[evalCheckSig](https://github.com/bitcoin/bitcoin/blob/a4ca4975880c4f870c6047065c70610af2529e74/src/script/interpreter.cpp#L391)’ does. This approach will allow for easier modification and extension for new versions of Silent addresses, and would also be more clear how version handling is done

josibake replied:
Thanks for the suggestion, I'll take a look at `EvalCheckSig`!

josibake replied:
Took a look at `EvalCheckSig` and while I think this is a good idea in general, I'd prefer to leave it as a follow-up for now. My reasoning here is we currently only have a `v0` silent payments version and I'd like to keep the code simple and focused on the `V0` rules. As a follow-up, or when we add support for a future `v1` version is likely a better time to introduce more complex code for handling multiple versions.

josibake commented on src/key.cpp:
```diff
@@ -172,6 +174,79 @@ bool CKey::Negate()
     return secp256k1_ec_seckey_negate(secp256k1_context_sign, keydata.data());
 }
 
+CKey CKey::AddTweak(const unsigned char *tweak32) const
```
rename to TweakAdd to more closely mirror the secp function

josibake commented on src/key.cpp:
```diff
@@ -172,6 +174,79 @@ bool CKey::Negate()
+
+// we dont want to hash the output of ECDH yet, so we provide a custom hash function to secp256k1_ecdh which
+// returns the serialized public key without hashing it
+int custom_ecdh_hash_function(unsigned char *output, const unsigned char *x32, const unsigned char *y32, void *data) {
```
this should also be rename since we aren't hashing

josibake replied:
move this to the UnhashedECDH function

josibake commented on src/key.cpp:
```diff
@@ -172,6 +174,79 @@ bool CKey::Negate()
+// This function returns the un-hashed ECDH result and is specific to silent payments.
+// If ECDH is required for a different use-case, a separate function should be used which does not use
+// the `custom_ecdh_hash_function` and instead returns the hashed ECDH result (as is standard practice)
+CPubKey CKey::SilentPaymentECDH(const CPubKey& pubkey) const
```
Should probably rename this to just "UnhashedECDH"

josibake commented on src/key_io.cpp:
```diff
@@ -309,3 +311,33 @@ bool IsValidDestinationString(const std::string& str)
+    return {scan_pubkey, spend_pubkey};
+}
+
+std::vector<unsigned char> DecodeSilentAddress(const std::string& str)
```
Similar to "Decode", there should be a higher-level function that first decodes the string and then passes it to the relevant decoding function

josibake replied:
Even better, should probably first add a new destination type to CTxDestination

josibake replied:
Resolved in #28246 , which adds a new `CTxDestination` and lets us use the existing encoder and decoder.

josibake commented on src/wallet/silentpayments.cpp:
```diff
@@ -0,0 +1,179 @@
+
+namespace wallet {
+
+Recipient::Recipient(const CKey& scan_seckey, const CKey& spend_seckey)
```
This doesn't need to be a struct, could just be functions which the DescScriptPubKeyMan calls as needed

achow101 replied:
https://github.com/achow101/bitcoin/commit/57d72b52ca5a442982907cb09c8b26e1ab5e9d59 can be squashed in here to remove the `Recipient` class.

josibake commented on src/wallet/silentpayments.cpp:
```diff
@@ -0,0 +1,179 @@
+    return {m_scan_pubkey, m_spend_pubkey};
+}
+
+std::vector<CKey> Recipient::ScanTxOutputs(std::vector<XOnlyPubKey> output_pub_keys)
```
this function should be "label aware," meaning if the wallet implements labels later on, this function does not need to change.

josibake replied:
Leaving this alone for now. This can be easily updated to add label support when it makes sense.

josibake commented on src/wallet/silentpayments.cpp:
```diff
@@ -0,0 +1,179 @@
+            return false;
+        }), output_pub_keys.end());
+    } while (!output_pub_keys.empty() && removed);
+    return raw_tr_keys;
```
This shouldn't return private keys, it should only return the tweak portion of the private key. This also means `Recipient` doesn't need to have the spend private key.

josibake commented on src/wallet/types.h:
```diff
@@ -62,6 +63,20 @@ enum class AddressPurpose {
+    bool fSubtractFeeFromAmount;
+};
+
+struct V0SilentPaymentDestination
```
this should be its own commit

josibake commented on src/wallet/wallet.h:
```diff
@@ -262,13 +262,6 @@ inline std::optional<AddressPurpose> PurposeFromString(std::string_view s)
     return {};
 }
 
-struct CRecipient
```
this should be its own commit, might not be needed if we instead V0SilentPaymentDestination as a CTxDestination type

Sjors commented:
Maybe point out in the description how the first commit 8283ca76e4feb37d3514fbd8dc491b8c4965dff1 relates to BIP 324, which doesn't use the ecdh module, but ElligatorSwift instead.

Sjors commented on src/key_io.cpp:
```diff
@@ -309,3 +311,33 @@ bool IsValidDestinationString(const std::string& str)
+    std::vector<unsigned char> scan_pubkey_data(data.begin(), data.begin() + 33);
+    CPubKey scan_pubkey{scan_pubkey_data};
+
+    std::vector<unsigned char> spend_pubkey_data(data.begin() + 33, data.end());
```
37b0ca1c3d753cc61e0debf8ea59160bdf4127ec: alternatively you could scan for the exact pubkey length. That would tolerate future versions to add extra stuff to the address that's safe to ignore for older versions.

josibake replied:
Done

Sjors commented on src/key_io.cpp:
```diff
@@ -309,3 +311,33 @@ bool IsValidDestinationString(const std::string& str)
+    if (!ConvertBits<5, 8, false>([&](unsigned char c) { silent_payment_data.push_back(c); }, dec.data.begin() + 1, dec.data.end())) {
+        return {};
+    }
+    if ((version == 0 && silent_payment_data.size() != SILENT_PAYMENT_V0_DATA_SIZE) || silent_payment_data.size() < SILENT_PAYMENT_V0_DATA_SIZE) return {};
```
37b0ca1c3d753cc61e0debf8ea59160bdf4127ec: could use `version == 0 && silent_payment_data.size() < SILENT_PAYMENT_V0_DATA_SIZE` (see comment above)

josibake replied:
I think this should fail if an address says it is v0 but has a data part larger than what is defined for v0. If an address has a version > v0 and data part > V0_SIZE, then only the _first_ 66 bytes of the data part are read, which is what we want for the forward compatibility bit.

Sjors commented on src/key_io.cpp:
```diff
@@ -309,3 +311,33 @@ bool IsValidDestinationString(const std::string& str)
     return IsValidDestinationString(str, Params());
 }
+
+std::pair<CPubKey, CPubKey> DecodeSilentData(const std::vector<unsigned char>& data)
```
37b0ca1c3d753cc61e0debf8ea59160bdf4127ec `assert data.size() >= SILENT_PAYMENT_V0_DATA_SIZE` (or early return if you don't expect the caller to have already checked this)

Sjors commented on src/key_io.cpp:
```diff
@@ -341,3 +341,21 @@ std::vector<unsigned char> DecodeSilentAddress(const std::string& str)
 }
+
+std::string EncodeSilentDestination(const CPubKey& scan_pubkey, const CPubKey& spend_pubkey)
+{
```
328951d056e939b6db573043fc71d8bd6ae68dff: if cheap, could assert/assume both ` CPubKey` are valid.

josibake replied:
This now uses the `src/key_io.cpp::Encode` function, which takes a `V0SilentPaymentDestination` object, so should be good. If we do want an assert, we can/should do it when creating the `V0SilentPaymentDestination` object

Sjors commented on src/key_io.cpp:
```diff
@@ -341,3 +341,21 @@ std::vector<unsigned char> DecodeSilentAddress(const std::string& str)
+    data_in.insert(data_in.end(), scan_pubkey.begin(), scan_pubkey.end());
+    data_in.insert(data_in.end(), spend_pubkey.begin(), spend_pubkey.end());
+
+    ConvertBits<8, 5, true>([&](unsigned char c) { data_out.push_back(c); }, data_in.begin(), data_in.end());
```
328951d056e939b6db573043fc71d8bd6ae68dff: this should be in some bech32 helper function.

josibake replied:
Not totally convinced we do want to wrap this in a helper function, but perhaps that can be a follow-up looking at how we handle bech32 across WitnessV0, WitnessV1 and V0SilentPaymentsDestinations

Sjors commented:
I think this PR should also contain `IsInputForSharedSecretDerivation` that's currently in the send PR, as well as `SumInputPubKeys` (not sure where that's introduced). More generally: anything that's critical in deriving the correct key from a transaction, given its inputs.

josibake committed:
Add V0SilentPaymentDestination address type

Sjors commented:
#28244 was merged

josibake commented:
> #28244 was merged

Awesome! I rebased on https://github.com/bitcoin/bitcoin/pull/28246, will respond to outstanding review comments today/tomorrow.

achow101 commented on src/wallet/silentpayments.h:
```diff
@@ -0,0 +1,49 @@
+
+    public:
+        Sender(const CKey& scalar_ecdh_input, const SilentPaymentRecipient& recipient);
+        std::vector<wallet::CRecipient> GenerateRecipientScriptPubKeys() const;
```
Please don't use `CRecipient` here, it causes a circular dependency.

josibake replied:
Fixed

josibake commented:
Fixed CI and fuzzing errors, will be responding to unaddressed review comments

josibake commented:
> In order to build the index PR #28241 directly on top of this and, more importantly, have more complete coverage of BIP 352 here, it would be useful to move `SumInputPubKeys` and `HashOutpoints` from the send / receive PR's here. These define which types of inputs are considered and how outpoints must be sorted and hashed.
> 
> (_update_: mmm, looking at the latest test vectors, maybe this functionality is already in here, just named differently)

I think the only thing missing now is `IsInputForSharedSecretDerivation`; I'll work on pulling that in next

josibake commented:
@hebasto can you rerun the Win64 CI task?

josibake commented on src/pubkey.h:
```diff
@@ -211,6 +211,12 @@ class CPubKey
     static bool CheckLowS(const std::vector<unsigned char>& vchSig);
 
+    //! Add a number of public keys together.
+    static CPubKey Combine(std::vector<CPubKey> pubkeys);
```
Fixed, and made pass by reference.

josibake commented on src/bech32.cpp:
```diff
@@ -370,11 +370,13 @@ std::string Encode(Encoding encoding, const std::string& hrp, const data& values
 /** Decode a Bech32 or Bech32m string. */
-DecodeResult Decode(const std::string& str) {
+DecodeResult Decode(const std::string& str, bool silent) {
+    std::size_t max_size = silent ? 1024 : 90;
```
This has been re-written to use a character limit enum.

josibake commented:
> In order to build the index PR #28241 directly on top of this and, more importantly, have more complete coverage of BIP 352 here, it would be useful to move `SumInputPubKeys` and `HashOutpoints` from the send / receive PR's here. These define which types of inputs are considered and how outpoints must be sorted and hashed.
> 
> (_update_: mmm, looking at the latest test vectors, maybe this functionality is already in here, just named differently)

This PR should have everything you need now. I've added a function `GetSilentPaymentTweakDataFromTxInputs`, which iterates over the transaction inputs, determines if they are eligible for silent payments and returns the sum of the input pubkeys, along with the hash of the tx outpoints.

The reason it returns the pubkey sum and outpoint hash separately is because if you are using this function to scan as a receiver, you want to calculate `outpoint_hash * scan_priv_key` first, and then do the ECDH step. If you're using this function to create an index, however, you'd want `sum_pubkeys * outpoint_hash` before storing it in the index.

achow101 commented on src/wallet/silentpayments.cpp:
```diff
@@ -0,0 +1,210 @@
+                size_t post_annex_size = txin.scriptWitness.stack.size() - (has_annex ? 1 : 0);
+                if (post_annex_size > 1) {
+                    // Actually a script path spend
+                    const std::vector<unsigned char>& control = txin.scriptWitness.stack.back();
```
In 644ff77be819a2dea15be0daad10fca8524a18f4 "Implement BIP352: Silent Payments"

Since we aren't modifying `txin.scriptWitness.stack` after checking for the annex, this may not actually be the control block. We need to remove the annex if it was there.

```suggestion
                    const std::vector<unsigned char>& control = txin.scriptWitness.stack.at(post_annex_size - 1);
```

josibake replied:
Done

achow101 commented on src/wallet/silentpayments.h:
```diff
@@ -0,0 +1,20 @@
+#include <undo.h>
+
+namespace wallet {
+CPubKey ComputeECDHSharedSecret(const CKey& scan_key, const CPubKey& sender_public_key, const uint256& outpoints_hash);
+uint256 ComputeTweak(const CPubKey& ecdh_pubkey, const uint32_t output_index);
+std::vector<uint256> GetTxOutputTweaks(const CPubKey& spend_pubkey, const CPubKey& ecdh_pubkey, std::vector<XOnlyPubKey> output_pub_keys);
+std::optional<std::pair<uint256, CPubKey>> GetSilentPaymentsTweakDataFromTxInputs(const std::vector<CTxIn>& vin, const std::map<COutPoint, Coin>& coins);
+
+CPubKey CreateOutput(const CKey& ecdh_scalar, const CPubKey& scan_pubkey, const CPubKey& spend_pubkey, const uint32_t output_index);
+uint256 HashOutpoints(const std::vector<COutPoint>& tx_outpoints);
+std::map<size_t, WitnessV1Taproot> GenerateSilentPaymentTaprootDestinations(const CKey& ecdh_scalar, const std::map<size_t, V0SilentPaymentDestination>& sp_dests);
+CKey PrepareScalarECDHInput(const std::vector<std::pair<CKey, bool>>& sender_secret_keys, const std::vector<COutPoint>& tx_outpoints);
```
In 644ff77be819a2dea15be0daad10fca8524a18f4 "Implement BIP352: Silent Payments"

It would be nice to have some documentation for these functions.

achow101 commented on src/wallet/silentpayments.cpp:
```diff
@@ -0,0 +1,210 @@
+#include <policy/policy.h>
+#include <script/solver.h>
+#include <util/check.h>
+#include <wallet/wallet.h>
```
In 644ff77be819a2dea15be0daad10fca8524a18f4 "Implement BIP352: Silent Payments"

`wallet/wallet.h` does not need to be included.

josibake replied:
Done

achow101 commented on src/wallet/test/silentpayment_tests.cpp:
```diff
@@ -0,0 +1,161 @@
+            const auto& expected = sender["expected"];
+
+            std::vector<COutPoint> outpoints;
+            for (const auto& outpoint : sender["given"]["outpoints"].getValues()) {
```
In 78d1426fabd7b5b6cdd28de508a1692fce291174 "Add BIP352 test vectors as unit tests"

nit: `s/sender["given"]/given`

```suggestion
            for (const auto& outpoint : given["outpoints"].getValues()) {
```

achow101 commented on src/wallet/test/silentpayment_tests.cpp:
```diff
@@ -0,0 +1,161 @@
+            const auto& expected = recipient["expected"];
+
+            std::vector<COutPoint> outpoints;
+            for (const auto& outpoint : recipient["given"]["outpoints"].getValues()) {
```
In 78d1426fabd7b5b6cdd28de508a1692fce291174 "Add BIP352 test vectors as unit tests"

nit: `s/recipient["given"]/given`

```suggestion
            for (const auto& outpoint : given["outpoints"].getValues()) {
````

josibake replied:
Done

achow101 commented on src/wallet/test/silentpayment_tests.cpp:
```diff
@@ -0,0 +1,161 @@
+            const auto expected_addresses = expected["addresses"].getValues();
+            // We know there is only one address, but if we support labels, this could be multiple addresses
+            CPubKey ecdh_pubkey = ComputeECDHSharedSecret(scan_priv_key, sum_input_pub_keys, HashOutpoints(outpoints));
+            std::vector<uint256> found_tweaks = GetTxOutputTweaks(spend_priv_key.GetPubKey(), ecdh_pubkey, output_pub_keys);
```
In 78d1426fabd7b5b6cdd28de508a1692fce291174 "Add BIP352 test vectors as unit tests"

nit: `s/spend_priv_key.GetPubKey()/spend_pubkey`

```suggestion
            std::vector<uint256> found_tweaks = GetTxOutputTweaks(spend_pubkey, ecdh_pubkey, output_pub_keys);
```

achow101 commented on src/wallet/wallet.h:
```diff
@@ -291,6 +291,11 @@ struct CRecipient
+
+    friend bool operator==(const CRecipient& a, const CRecipient& b)
+    {
+        return a.dest == b.dest && a.nAmount == b.nAmount;
```
In 78d1426fabd7b5b6cdd28de508a1692fce291174 "Add BIP352 test vectors as unit tests"

Shouldn't this also check sffo too?

```suggestion
        return a.dest == b.dest && a.nAmount == b.nAmount && a.fSubtractFeeFromAmount == b.SubtractFeeFromAmount;
```

josibake replied:
Maybe? If we want to know if the actual objects themselves are equal, then yes, but if we wan't a more conceptual "are these the same transaction output," then I'd argue no?

achow101 commented on src/wallet/test/silentpayment_tests.cpp:
```diff
@@ -0,0 +1,161 @@
+
+CKey GetKeyFromBIP32Path(std::vector<std::byte> seed, std::vector<uint32_t> path)
+{
+    CExtKey key_parent, key_child;
+    key_parent.SetSeed(seed);
+    for (auto index : path) {
+        BOOST_CHECK(key_parent.Derive(key_child, index));
+        std::swap(key_parent, key_child);
+    }
+    return key_parent.key;
```
In 78d1426fabd7b5b6cdd28de508a1692fce291174 "Add BIP352 test vectors as unit tests"

No need to have two `CExtKey`s and swapping them, you can just derive a key into itself.

```suggestion
    CExtKey key;
    key.SetSeed(seed);
    for (auto index : path) {
        BOOST_CHECK(key.Derive(key, index));
    }
    return key.key;
```

josibake replied:
Done

achow101 commented:
Currently, this PR is dangerous by itself as it will accept silent payment addresses for sending, but fail to actually send the coins to the correct script. It just uses an empty script. We should disallow sending to silent payment addresses in this PR, and enable it in the sending PR.

josibake commented on src/key_io.cpp:
```diff
@@ -127,12 +146,29 @@ CTxDestination DecodeDestination(const std::string& str, const CChainParams& par
+                return CNoDestination();
+            }
+            auto version = dec.data[0];  // retrieve the version
+            if (version == 0 && data.size() != SILENT_PAYMENT_V0_DATA_SIZE) {
+                error_str = strprintf("Silent payment version is v0 but data is not the correct size (expected %d, got %d).", SILENT_PAYMENT_V0_DATA_SIZE, data.size());
+                return CNoDestination();
+            }
```
Since SP addresses are forward-compatible, we don't care: we always read the first 66 bytes as two pubkeys and ignore the rest. However, we do need a check here for v31 or greater, as this indicates a backward incompatible change

josibake replied:
Added a check for version >= 31

josibake commented on src/util/message.cpp:
```diff
@@ -47,7 +47,7 @@ MessageVerificationResult MessageVerify(
     }
 
-    if (!(CTxDestination(PKHash(pubkey)) == destination)) {
+    if (!(PKHash(pubkey) == *std::get_if<PKHash>(&destination))) {
```
This is a fix for an `-Wuninitialized` warning, see https://github.com/bitcoin/bitcoin/pull/28246#discussion_r1318874324

josibake commented on src/addresstype.h:
```diff
@@ -109,9 +117,14 @@ using CTxDestination = std::variant<CNoDestination, PKHash, ScriptHash, WitnessV
+ *
+ * For standard scripts that have addresses (and P2PK as an exception), a corresponding CTxDestination
+ * is assigned to addressRet.
+ * For all other scripts. addressRet is assigned as a CNoDestination containing the scriptPubKey.
```
These commits belong to https://github.com/bitcoin/bitcoin/pull/28246 , might be better to leave a review there

theStack replied:
Oops, indeed! It might be a good idea to read PR descriptions before starting a review 😅

josibake commented on src/key.cpp:
```diff
@@ -172,6 +174,62 @@ bool CKey::Negate()
+CPubKey CKey::UnhashedECDH(const CPubKey& pubkey) const
+{
+    unsigned char shared_secret[33];
+
+    secp256k1_pubkey ecdh_pubkey;
```
Hm, it seems a bit odd to assert that the passed pubkey is valid? I think it's safe to assume that validation was done upstream by the caller. Also, I agree that we want to avoid doing expensive checks here as this function is called during scanning.

I'll add the assert for the private key, though, as this seems in keeping with what the other code with private keys is doing.

josibake replied:
Added `assert(fValid)` check for CKey and also made the functions consistent with the rest of the CPubKey/CKey functions by returning a bool instead of an `assert` inside a void function

josibake commented on src/pubkey.h:
```diff
@@ -287,10 +294,19 @@ class XOnlyPubKey
     bool operator!=(const XOnlyPubKey& other) const { return m_keydata != other.m_keydata; }
     bool operator<(const XOnlyPubKey& other) const { return m_keydata < other.m_keydata; }
 
+    CPubKey ConvertToCompressedPubKey(bool even = true) const
```
Good catch! This is dead code from an old version of the PR, will remove it

josibake replied:
Fixed

josibake committed:
wallet: disable sending to silent payment address

Have `IsValidDestination` return false for silent payment destinations
and set an error string when decoding a silent payment address.

This prevents anyone from sending to a silent payment address before
sending is implemented in the wallet, but also allows the functions to
be used in the unit testing famework.

josibake commented on src/key_io.cpp:
```diff
@@ -64,6 +66,21 @@ class DestinationEncoder
+        data_in.reserve(66);
+        // Set 0 as the silent payments version
+        std::vector<unsigned char> data_out = {0};
+        data_out.reserve(67);
```
Fixed and added a comment

josibake commented on src/pubkey.cpp:
```diff
@@ -256,6 +259,22 @@ std::optional<std::pair<XOnlyPubKey, bool>> XOnlyPubKey::CreateTapTweak(const ui
+    return_val = secp256k1_ec_pubkey_serialize(secp256k1_context_static, pubkey_bytes, &publen, &original_pubkey, SECP256K1_EC_COMPRESSED);
+    assert(return_val);
+
+    return CPubKey(pubkey_bytes);
```
Fixed, and also updated CKey::TweakAdd,TweakMultiply to return a bool (instead of an assert + void), as this better matches the CKey/CPubKey api

josibake commented:
`git range-diff 0cf23d0...d053c3a`

Addressed review comments from @theStack and @achow101 and added a commit to disable sending to silent payment addresses, while still allowing the functions to be used in the unit tests.

josibake commented:
> Currently, this PR is dangerous by itself as it will accept silent payment addresses for sending, but fail to actually send the coins to the correct script. It just uses an empty script. We should disallow sending to silent payment addresses in this PR, and enable it in the sending PR.

Fixed in https://github.com/bitcoin/bitcoin/pull/28122/commits/d053c3af90324df4fddc57c8301fcf414019d726

josibake commented:
Next step: add documentation for the new functions, requested in https://github.com/bitcoin/bitcoin/pull/28122#discussion_r1321966435

achow101 commented on src/wallet/silentpayments.cpp:
```diff
@@ -0,0 +1,241 @@
+{
+    // Scan first for any outputs to us, and then check if there is any change
+    std::vector<uint256> tweaks = GetTxOutputTweaks(spend_pubkey, ecdh_pubkey, output_pub_keys);
+    std::vector<uint256> labeled_tweaks = GetTxOutputTweaks(change_label.second, ecdh_pubkey, output_pub_keys);
```
In 93ff4ce117d4b85128dd55630d2dc9f8efbfb473 "Implement BIP352: Silent Payments"

In the current receiving PR, `change_label` is a pair of the change label and it's corresponding curve point. However this is expecting it as the pair of the change label and the resulting spend pubkey.

josibake commented:
`git range-diff  c9dff7e...5829940` changes `/*compressed=*/` to `/*fCompressedIn=*/` to fix a tidy warning

josibake commented:
Pulled in support for scanning with labels (h/t @achow101) and updated the unit tests. Major changes:

* Add new functions for generating a change label and labeled silent payments address, per BIP352
* Update `GetTxOutputTweaks` to scan with labels
* Refactor `CPubKey` to get rid of `Combine` in favor of `operator+`

Ready for another round of review!

Sjors commented:
@josibake can you add the following helper functions?

- [ ] `CouldBeSilentPayment(CTransactionRef tx)`: checks that it's not a coinbase and that at least one of the outputs is taproot
See https://github.com/bitcoin/bitcoin/pull/28241#discussion_r1331881346
- [ ] `bool GetSilentPaymentsTweakedPubKeySum(std::vector<CTxIn>, std::map<COutPoint, Coin>, CPubKey& tweaked_key) nodiscard` (with clear documentation on what to do when it returns `false`). Although relatively straight-forward, my current index implementation does look a bit brittle, and I'm not sure how to handle failure cases: https://github.com/bitcoin/bitcoin/pull/28241/files#diff-708eeba963ec57793dbec2309d27f32869d4ccefa14fbeb575d3a1e36e5e1a03R74-R85

josibake commented:
force-pushed to fix a silent merge conflict with #28500

S3RK commented on src/wallet/test/silentpayment_tests.cpp:
```diff
@@ -0,0 +1,172 @@
+            }
+
+            std::vector<std::pair<CKey, bool>> sender_secret_keys;
+            for (const auto& key : given["input_priv_keys"].getValues()) {
```
With current implementation, we can't have test vectors for sending that include non-eligible sp inputs. This is because we just sum all the private keys without checking if the input is eligible. I propose to extract eligibility checking as a function from `GetSilentPaymentsTweakDataFromTxInputs` so we can use it here in the unit test.

S3RK commented on src/wallet/silentpayments.cpp:
```diff
@@ -0,0 +1,294 @@
+    return tweaks;
+}
+
+std::optional<std::pair<uint256, CPubKey>> GetSilentPaymentsTweakDataFromTxInputs(const std::vector<CTxIn>& vin, const std::map<COutPoint, Coin>& coins)
```
According to the BIP we should skip the whole tx if there's unknown witness input. As currently, implemented it will silently skip just that input and attempt to proceed. I added a test vector for unknown witness version

josibake commented:
rebased on master and updated to fix a silent merge conflict in the tests, related to the new `Txid` type.

josibake commented:
> @josibake can you add the following helper functions?
> 
>     * [ ]  `CouldBeSilentPayment(CTransactionRef tx)`: checks that it's not a coinbase and that at least one of the outputs is taproot
>       See [Silent payment index (for light wallets and consistency check) #28241 (comment)](https://github.com/bitcoin/bitcoin/pull/28241#discussion_r1331881346)
> 
>     * [ ]  `bool GetSilentPaymentsTweakedPubKeySum(std::vector<CTxIn>, std::map<COutPoint, Coin>, CPubKey& tweaked_key) nodiscard` (with clear documentation on what to do when it returns `false`). Although relatively straight-forward, my current index implementation does look a bit brittle, and I'm not sure how to handle failure cases: https://github.com/bitcoin/bitcoin/pull/28241/files#diff-708eeba963ec57793dbec2309d27f32869d4ccefa14fbeb575d3a1e36e5e1a03R74-R85
> 
> 
> Maybe also make a type alias for `std::map<COutPoint, Coin>`

the second function you're asking for exists, just need to expose it in the header file. it also returns `pair<uint256, CPubKey>`, so you would need to multiply those together (using the TweakMultiply method on CPubKey) before storing in the index.

For the first function, we can probably roll that into the second one? Need to look at the code again to see if that makes sense.

Regarding the alias, I'm somewhat ambivalent. Since we use this same struct in other places unrelated to silent payments, it might make sense as a follow-up refactor. Less inclined to do it here.

josibake commented:
Rebased https://github.com/bitcoin/bitcoin/commit/0299dfb0607b85813c7226690772ab3be198b0df -> https://github.com/bitcoin/bitcoin/commit/6a9ec4771d691bf4143f3574dd129da641c6ee74

* Fixes a silent merge conflict introduced by changing `CKey` to use `std::byte*` instead of `unsigned char`

josibake commented:
Updated https://github.com/bitcoin/bitcoin/commit/6a9ec4771d691bf4143f3574dd129da641c6ee74 -> https://github.com/bitcoin/bitcoin/pull/28122/commits/6e7c29528eb4a7c39bf8290b7c622e30bf07b45d ([implement-bip352-00](https://github.com/josibake/bitcoin/tree/implement-bip352-00) -> [implement-bip352-01](https://github.com/josibake/bitcoin/tree/implement-bip352-01), [compare](https://github.com/josibake/bitcoin/compare/implement-bip352-00..implement-bip352-01))
* Adds a test case for taproot inputs with H as the internal key (with and without the optional annex) (h/t @Eunovo)
* Splits out the pubkey extraction logic from `GetSilentPaymentTweakDataFromTxInputs` function into its own function: `ExtractPubKeyFromInput`
* Updates the sending tests to check if the input is a silent payments input before using the secret key, using the `ExtractPubKeyFromInput` function

Going forward, it might make more sense to have a function that tests if the input is eligible (`bool IsInputForSharedSecretDerivation`, and a separate function to extract the pubkey. Both sending and receiving need the `bool IsInputForSharedSecretDerivation` function, but only the receiver would need the pubkey extraction function. Right now, it's using the extract function for both by checking if a pubkey was extracted or not, which seems kinda clunky but works for now.

Eunovo commented on src/wallet/silentpayments.cpp:
```diff
@@ -0,0 +1,294 @@
+                // Check if the redeemScript is P2WPKH
+                CScript redeem_script{stack.back().begin(), stack.back().end()};
+                TxoutType rs_type = Solver(redeem_script, solutions);
+                if (rs_type == TxoutType::WITNESS_V0_KEYHASH) {
```
@josibake p2sh-p2pkh or p2sh-p2pk are not addressed in the BIP. While I understand they are not commonly used scripts but `Solver(redeem_script, solutions)` can still detect p2pkh or p2pk in the redeem_script. Is there anything against adding support for these scripts? assuming that we ignore non-standard or malleated p2pkh in the redeem script

josibake replied:
The goal is not to support every possible script. Rather, we want to limit it to a sane subset to avoid technical debt and keep the implementation simple. In fact, the only reason for including `P2PKH` at all is due to the fact it is still widely used and makes up a significant percentage of the UTXO set, otherwise we would have only allowed Segwit script templates as this gives us non-malleability guarantees.

Eunovo replied:
Understood. Thanks for clearing that up @josibake

josibake commented:
Rebased https://github.com/bitcoin/bitcoin/commit/e0ea52da97b5bb18b813e862455aafd56a7809dd -> https://github.com/bitcoin/bitcoin/commit/08b62764190a907c0d7d329f4df07f9a2b6dcdeb 

* Fixed small `configure.ac` merge conflict

josibake commented:
Updated https://github.com/bitcoin/bitcoin/commit/08b62764190a907c0d7d329f4df07f9a2b6dcdeb -> https://github.com/bitcoin/bitcoin/commit/cbb8d3a6574e10841c52716bbcd27e0c60752e1c ([implement-bip352-01-rebase](https://github.com/josibake/bitcoin/tree/implement-bip352-01-rebase) -> [implement-bip352-02](https://github.com/josibake/bitcoin/tree/implement-bip352-02) , [compare](https://github.com/josibake/bitcoin/compare/implement-bip352-01-rebase..implement-bip352-02))

* Added doxygen style comments to `src/wallet/silentpayments.h`
* Made `GetTxOutputTweaks` return a `std::optional` to cover the case where no payments to us are found

S3RK commented on src/wallet/silentpayments.h:
```diff
@@ -0,0 +1,178 @@
+ *
+ * Given a set of silent payment destinations, generate the requested number of outputs. If a silent payment
+ * destination is repeated, this indicates multiple outputs are requested for the same recipient. The silent payment
+ * desintaions are passed in map where the key indicates their desired position in the final tx.vout array.
```
Not sure protocol implementation needs to now about the position in tx.vout array. AFAIK, output calculation doesn't depend on the position. Have you considered passing destinations one by one together with the counter if a destination is repeated?

josibake replied:
Output calculation doesn't depend on the position, but using output position makes the implementation much simpler. I did try something similar to what you suggested but it ended up being really complicated to make sure I was matching up all the right amounts and was always _implicitly_ relying on output position. @achow101 pointed out it would be better and simpler to just use the output index explicitly.

This also doesn't create any dependence on output position. The user can still order the outputs however they want after generating the silent payment scriptPubKeys.

S3RK commented on src/wallet/silentpayments.h:
```diff
@@ -0,0 +1,178 @@
+ * Get the necessary data from the transaction inputs to be able to scan the transaction outputs for silent payment outputs.
+ * This requires knowledge of the prevout scriptPubKey, which is passed via `coins`.
+ *
+ * If the transaction has silent payment eligible inputs, the public keys from these inputs are returned as a sum along with
```
I'm too dumb to understand why do we need to return sum on input pubkeys together with the input hash. Does it have something to do with index and light clients?

josibake replied:
Yep! There are two ways to use this data:

1. Immediately, as `(b_scan * input_hash) * sum_input_pubkeys`. This is one scalar multiplication and one EC multiplication.
2. Store it for later (either for wallet re-scans or to serve to light clients), as `input_hash * sum_input_pubkeys`. This is one EC multiplication and allows us to store all of the input data needed as 33 bytes.

If the function instead returned `input_hash * sum_input_pubkeys`, we would end up doing two EC multiplications for case 1 (i.e. `(input_hash * sum_input_pubkeys) * b_scan`)

S3RK reviewed (commented):
Started reviewing. Basic question first. Which RPCs are supposed to work at this stage? I was under impression, that none RPCs should understand SP addresses yet and the support come with send and receiving PRs. But since you've registered SP destination in `CTXDestination` variant, the SP addresses would be recognised by some RPCs. Is that intended? Should we verify that they return sane results?

josibake commented:
> Started reviewing. Basic question first. Which RPCs are supposed to work at this stage? I was under impression, that none RPCs should understand SP addresses yet and the support come with send and receiving PRs. 

Thanks for the review @S3RK ! The goal isn't that none of the RPCs should understand SP addresses, rather the goal is that _full_ RPC support is deferred to the send and receiving PRs.

> But since you've registered SP destination in `CTXDestination` variant, the SP addresses would be recognised by some RPCs. Is that intended? Should we verify that they return sane results?

In "wallet: disable sending to silent payment address" https://github.com/bitcoin/bitcoin/pull/28122/commits/ed75ce5362ab69c5968b85e42c45683519325626 SP addresses are marked as invalid by the `ValidDestinationVisitor` (which should cover all of the RPCs where an address can be passed) and `validateaddress` returns a message that this is a silent payment address, but support is not implemented yet.

josibake committed:
common: add bip352.{h,cpp} secp256k1 module

Wrap the silentpayments module from libsecp256k1. This is placed in
common as it is intended to be used by:

  * RPCs: for parsing addresses
  * Wallet: for sending, receiving, spending silent payment outputs
  * Node: for creating silent payment indexes for light clients

josibake commented:
I've updated this to use https://github.com/bitcoin-core/secp256k1/pull/1471 for the cryptography needed for silent payments. This way, we don't need make any changes to `CKey`/`CPubKey` and keeps all of the low-level cryptography in `libsecp256k1`. This simplifies this PR to creating the necessary wrappers for the libsecp silent payments module and implementing the non-cryptography parts of the protocol (i.e. the client code).

Putting this PR in draft for now, until https://github.com/bitcoin-core/secp256k1/pull/1471 merges.

Sjors commented on src/wallet/silentpayments.h:
```diff
@@ -0,0 +1,93 @@
+ * @param coins                                            The coins (potentially) spent in this transaction.
+ * @return std::optional<std::pair<uint256, PubTweakData>> The silent payment tweak data, or nullopt if not found.
+ */
+std::optional<BIP352::PubTweakData> GetSilentPaymentTweakDataFromTxInputs(const std::vector<CTxIn>& vin, const std::map<COutPoint, Coin>& coins);
```
Assuming I still need this for the index #28241, I still prefer if this wasn't part of the wallet. Afaik nothing in this file has a dependency on the wallet, so the easiest change would to be move this to `src/common`.

josibake replied:
Now that we have `src/bip352.h`, it might make more sense to get rid of `src/wallet/silentpayents.h` and put everything in `src/bip352.h`. In the future, anything that is wallet specific can go in `src/wallet/silentpayments.h` and you'd have everything you need for an index in `src/bip352.h`. Thoughts?

Sjors commented on src/primitives/transaction.h:
```diff
@@ -43,7 +44,10 @@ class COutPoint
+        DataStream ser_a, ser_b;
+        ser_a << a;
+        ser_b << b;
+        return Span{ser_a} < Span{ser_b};
```
7a27533332d14cf98f7e165f464142132e9f9d85: if you're going to change this here instead of adding a custom sort for BIP352, then you should add a note that it's critical for BIP352.

However I'm not sure if we should sort this way by default. If I wanted to list transaction outputs, I'd want to sorted by their integer version.

That said, I don't know if we currently expose any RPC where this sort order is exposed.

josibake replied:
From what I could tell, we don’t expose this anywhere. I’m in favor of making this the default since we only ever really talking about outpoints in terms of their serialised encoding. Gonna give it some more thought, but I’m considering opening this commit as its own PR. It seems preferable to me to have one way of sorting outpoints vs creating a special case for silent payments, especially when there doesn’t seem to be any other “sort outpoints” use case (e.g. none of the tests failed when I made this change).

josibake replied:
But still need to verify re: displayed in RPCs. At that point tho, we could just sort by `vout` and ignore the txid in the sort

Sjors replied:
Separate PR sounds good to get more eyes on it. Meanwhile this is fine a temporary fix for this PR.

josibake replied:
Closed #30046 :smile:

Sjors commented:
@josibake can you add a link to https://github.com/bitcoin-core/secp256k1/pull/1519 in the PR description? It currently takes a few "load more" clicks and searching to see what this PR depends on.

josibake committed:
tests: add BIP352 test vectors as unit tests

Use the test vectors to test sending and receiving. A few cases are not
covered here, namely anything that requires testing specific to the
wallet. For example:

* Taproot script path spending is not tested, as that is better tested in
  a wallets coin selection / signing logic
* Re-computing outputs during RBF is not tested, as that is better
  tested in a wallets RBF logic

The unit tests are written in such a way that adding new test cases is
as easy as updating the JSON file

Sjors commented:
Now might be a good time to rebase this and the send and receive branches? The "pre-work" commits mentioned in the description are merged and the libsecp PR seems to be in reasonable shape.

Sjors commented:
I think you lost https://github.com/bitcoin/bitcoin/pull/30046/commits/e11d764a6ce72cd571ad9cf818c6918dd16b02f6 in the rebase?

josibake commented:
Regarding the @DrahtBot comment, I am still working on this PR and will be rebasing this week, once I've finished responding to review feedback in https://github.com/bitcoin-core/secp256k1/pull/1519

josibake committed:
crypto: add read-only method to KeyPair

Add a method for passing a KeyPair object to secp256k1 functions expecting a secp256k1_keypair.
This allows for passing a KeyPair directly to a secp256k1 function without needing to create a
temporary secp256k1_keypair object.

josibake commented:
CI failure is due to the "test each commit" job not being able to handle a subtree merge commit. Ignoring for now; will look into fixing the CI job.

Eunovo commented on src/common/bip352.cpp:
```diff
@@ -0,0 +1,384 @@
+
+        pubkey = XOnlyPubKey{solutions[0]};
+    } else if (type == TxoutType::WITNESS_V0_KEYHASH) {
+        pubkey = CPubKey{txin.scriptWitness.stack.back()};
```
https://github.com/bitcoin/bitcoin/pull/28122/commits/87a6ea213301342afbdd37876939c1eeb142b8e3: This will cause a segfault when `scriptWitness` is empty. `scriptWitness` can be empty if the tranasaction hasn't been signed yet. @macgyver reached this edge case while testing #32966 see https://github.com/bitcoin/bitcoin/pull/32966#discussion_r2277212074

josibake committed:
Squashed 'src/secp256k1/' changes from b9313c6e1a..aa85bfb530

aa85bfb530 docs: update README
9f42a30b82 ci: enable silentpayments module
d504e48145 tests: add sha256 tag test
124750d580 tests: add constant time tests
b35ffa2e30 tests: add BIP-352 test vectors
038c5b9c9d silentpayments: add benchmarks for scanning
88eb3d4545 silentpayments: add examples/silentpayments.c
22b20fd617 silentpayments: receiving
df1de93765 silentpayments: recipient label support
76a0451c76 silentpayments: sending
3cd3a93bff build: add skeleton for new silentpayments (BIP352) module
f36afb8b3d Merge bitcoin-core/secp256k1#1725: tests: refactor tagged hash verification
5153cf1c91 tests: refactor tagged hash tests
d2dcf52091 Merge bitcoin-core/secp256k1#1726: docs: fix broken link to Tromer's cache.pdf paper
489a43d1bf docs: fix broken link to eprint cache.pdf paper
d599714147 Merge bitcoin-core/secp256k1#1722: docs: Exclude modules' `bench_impl.h` headers from coverage report
0458def51e doc: Add `--gcov-ignore-parse-errors=all` option to `gcovr` invocations
1aecce5936 doc: Add `--merge-mode-functions=separate` option to `gcovr` invocations
106a7cbf41 doc: Exclude modules' `bench_impl.h` headers from coverage report
a9e955d3ea autotools, docs: Adjust help string for `--enable-coverage` option
e523e4f90e Merge bitcoin-core/secp256k1#1720: chore(ci): Fix typo in Dockerfile comment
24ba8ff168 chore(ci): Fix typo in Dockerfile comment
74b8068c5d Merge bitcoin-core/secp256k1#1717: test: update wycheproof test vectors
c25c3c8a88 test: update wycheproof test vectors
20e3b44746 Merge bitcoin-core/secp256k1#1688: cmake: Avoid contaminating parent project's cache with `BUILD_SHARED_LIBS`
2c076d907a Merge bitcoin-core/secp256k1#1711: tests: update Wycheproof
7b07b22957 cmake: Avoid contaminating parent project's cache with BUILD_SHARED_LIBS
5433648ca0 Fix typos and spellings
9ea54c69b7 tests: update Wycheproof files

git-subtree-dir: src/secp256k1
git-subtree-split: aa85bfb530b9ffc3dde6eaa7a976e129b8bd2f58

josibake committed:
Merge commit '05c5447f2a9c1db7f7dc210b24d2e37c28af0c1a' into refresh-secp256k1
